R/functions.R

```diff
@@ -100,55 +62,166 @@ skim_with_defaults <- function() {
-#
-# The call to purrr::detect returns the first appropriate match for objects
-# that have more than one class. Doing it this way saves us from having
-# to define methods for multiple classes.
-#
-# @param type The type of summary functions to extract
-# @return A list of summary functions
-# @keywords internal
-# @export
+#' An internal method for getting a set of summary functions, by type. We use
+#' this approach instead of method dispatch because we want to be able to
+#' dynamically add or remove the summary functions for each type.
+#'
+#' The call to purrr::detect returns the first appropriate match for objects
+#' that have more than one class. Doing it this way saves us from having
+#' to define methods for multiple classes.
+#'
+#' @param type The type of summary functions to extract
+#' @return A list of summary functions
+#' @keywords internal
+#' @noRd
 
 get_funs <- function(type) {
-  all <- functions$current[type]
+  all <- options$functions[type]
   purrr::detect(all, purrr::compose(`!`, is.null))
 }
 
 
-# A method for getting the name of set of summary functions names, by type.
-# This function is similar to get_funs, in that it applies a type of dispatch
-# for the type of object provided. This function gets the name of the
-# group of summary functions for the type. This lets the user know what set of
-# functions are used in producing the skim_df.
-#
-# @param type The type of summary functions to extract
-# @return A length-one character vector that shows the class that was matched
-#  by skimr.
-# @keywords internal
-# @export
-
-get_fun_names <- function(type) {
-  all <- functions$current[type]
+#' A method for getting the name of set of summary functions names, by type.
+#' This function is similar to get_funs, in that it applies a type of dispatch
+#' for the type of object provided. This function gets the name of the
+#' group of summary functions for the type. This lets the user know what set of
+#' functions are used in producing the skim_df.
+#'
+#' @param type The type of summary functions to extract
+#' @return A length-one character vector that shows the class that was matched
+#'  by skimr.
+#' @keywords internal
+#' @noRd
+
+get_vector_type_used <- function(type) {
+  all <- options$functions[type]
   id <- purrr::detect_index(all, purrr::compose(`!`, is.null))
   names(all)[id]
-}
\ No newline at end of file
+}
+
+
+# Default summarizing functions for each type -----------------------------
+
+numeric_funs <- list(
+  missing = n_missing,
+  complete = n_complete,
+  n = length,
+  mean = purrr::partial(mean, na.rm = TRUE),
+  sd = purrr::partial(sd, na.rm = TRUE),
+  min = purrr::partial(min, na.rm = TRUE),
+  p25 = purrr::partial(quantile, probs = .25, na.rm = TRUE, names = FALSE),
+  median = purrr::partial(median, na.rm = TRUE),
+  p75 = purrr::partial(quantile, probs = .75, na.rm = TRUE, names = FALSE),
+  max = purrr::partial(max, na.rm = TRUE),
+  hist = inline_hist
```

Comment:
What are the performance implications of going to quantile twice as opposed to once to get all three (or even all 5)?

Reply:
Haven't thought about it TBH. My bigger concern was solving the issues with output.

If a function produces a single value, skim adds it as a column. If a functions produces multiple values, skim now produces name: value pairs. I think this looks really nice for factors, which don't have a high impact graphic like the histogram.

Reply:
That's the actual issue with median and why I was thinking that we switch to calling it 50th percentile.  Also I want to deal with time series and that (but that's a small issue).